### DISCUSSION TITLE
Does efficiency of casadi code affect the final performance of the solver?

### DISCUSSION THREAD

**Discussion by sandeep026** [alphaville/optimization-engine#328] [Q&A] [answered]
Having worked in casadi a lot, the complexity of the expression graph can depend on the person coding for a given application. I think this can affect the overall performance of the solver when it comes to function evaluations. For example, multiple shooting methods can be evaluated in serial and parallel.

I would like your suggestions before working on OpEn.

**Comment by alphaville** [comment, marked as answer]
@sandeep026 We use CasADi to generate C code for some gradients. The C code is then co-compiled with our Rust code. The generated code can indeed be very lengthy. We have a [paper](https://arxiv.org/pdf/2306.10861) where we implemented a custom code generator for optimal control problems, which has a significantly lower footprint in terms of geneated code:

![image](https://github.com/user-attachments/assets/9e18f7b1-90ed-443c-9773-eb6cf9c5263e)

and in some examples seems to run faster (esp. for large prediction horizons):

![image](https://github.com/user-attachments/assets/903b5568-466e-46b6-bb72-c6056b98f9fd)

We have implemented this in [`gradgen`](https://github.com/QUB-ASL/gradgen), but we haven't integrated it with OpEn. 

That said, overall I find that CasADi performs decently well, despite the fact that it can lead to very large auto-generated files.
